File: src/main/java/hashmaps/RansomNote.java
package hashmaps;

import java.util.HashMap;
import java.util.Map;

public class RansomNote {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> magazineCount = new HashMap<>();

        for (char c : magazine.toCharArray()) {
            magazineCount.put(c, magazineCount.getOrDefault(c, 0) + 1);
        }
        for (char c : ransomNote.toCharArray()) {
            if (magazineCount.getOrDefault(c, 0) != 0) {
                magazineCount.put(c, magazineCount.getOrDefault(c, 0) - 1);
            } else {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {

    }
}

File: leet-code-Go/hashmaps/Ransom_Note.go
package hashmaps

type RansomNote struct{}

func CanConstruct(ransomNote string, magazine string) bool {
	cnt := [26]int{}
	for _, c := range magazine {
		cnt[c-'a']++
	}
	for _, c := range ransomNote {
		cnt[c-'a']--
		if cnt[c-'a'] < 0 {
			return false
		}
	}
	return true
}

func CanConstructMap(ransomNote string, magazine string) bool {
	cnt := make(map[rune]int)
	for _, c := range magazine {
		cnt[c]++
	}
	for _, c := range ransomNote {
		cnt[c]--
		if cnt[c] < 0 {
			return false
		}
	}
	return true
}
